### ISSUE TITLE
Fix project renaming issues - renaming project can leave global config with two projects

### ISSUE THREAD

**Issue by wolffc** [drud/ddev#1933]
**Describe the bug**
if you rename your project ddev seems to ad a new entry for the same project over and over again

Updated: If you edit an existing configured project and change its name, ddev will let you start that; it shouldn't, just as it doesn't allow the same name to be used.

**To Reproduce**
1. create a new project `ddev config` everything default 
2. run the following script (to change the project name multiple times)
```bash
#!/usr/bin/env bash

# enable error reporting and show executed commands
set -ex

changename()
{
  # stop to have a clean state
  ddev stop
  # reconfigure project name
  echo "name: $1" > .ddev/config.site.yaml
  # restart the project with the new name
  ddev start;
}

changename "foo-example"
changename "bar-example"
changename "foo-example"
changename "bar-example"
```
3 `ddev list` shows somthing like
```
bar-example             php    C:/src/ddev-namechange         https://bar-example.ddev.site     running
bar-example             php    C:/src/ddev-namechange         https://bar-example.ddev.site     running
bar-example             php    C:/src/ddev-namechange         https://bar-example.ddev.site     running
```

**Expected behavior**
ddev should not dublicate the project every time it is renamed. 

**Version and configuration information (please complete the following information):**
 - Windows 10 (19.17763 Build 17763)
- Docker
 docker version
Client: Docker Engine - Community
 Version:           19.03.4
 API version:       1.40
 Go version:        go1.12.10
 Git commit:        9013bf5
 Built:             Thu Oct 17 23:44:48 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.4
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.10
  Git commit:       9013bf5
  Built:            Thu Oct 17 23:50:38 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
 

 - ddev  v1.11.2
 - config.yaml contents for the misbehaving project
```
APIVersion: v1.11.2
name: foo-bar
type: php
docroot: ""
php_version: "7.2"
webserver_type: nginx-fpm
router_http_port: "80"
router_https_port: "443"
xdebug_enabled: false
additional_hostnames: []
additional_fqdns: []
mariadb_version: "10.2"
nfs_mount_enabled: false
provider: default
use_dns_when_possible: true
timezone: ""
```
 - just the config.site.yaml which is generated by the above script. 

**Additional context**
we have a Polymorph Project (one project we use to deploy multiple sites). we rename the project so docker keeps the old container data an we are able to fast-switch between different sites.

**Comment by Graloth** [comment]
I can confirm this issue, can replicate it as per the issue description, but initially encountered it when adding the same hostname twice by mistake.

An addition to this issue is that once I've removed the duplicated hostnames, it continues to appear multiple times in `ddev list` both when the project is stopped and started.

**Comment by rfay** [comment]
Could you folks please provide the contents of the project_info stanza of ~/.ddev/global_config.yaml when this happens? Thanks.

@wolffc I expect that your polymorph project does violate the concept of unique project names on the machine... There can be only one ddev project of a given name on a machine.

**Comment by Graloth** [comment]
@rfay I'm not sure what you mean, I do not have any `global_config.yaml` inside my .ddev directory, I assume it's just the default `config.yaml` that is initially generated by DDEV.

If that is the case, this is mine

```yaml
APIVersion: v1.11.2
name: foo-bar
type: drupal8
docroot: web
php_version: "7.2"
webserver_type: nginx-fpm
router_http_port: "80"
router_https_port: "443"
xdebug_enabled: false
project_tld: dk.localhost
additional_hostnames: []
additional_fqdns: []
mariadb_version: "10.2"
nfs_mount_enabled: false
provider: default
use_dns_when_possible: false
timezone: "Europe/Copenhagen"
```

This is a clean installed test project (Drupal in my case). The only changes was to add timezone, turn off use dns and add a custom 

I haven't been able to replicate it again manually, only using the bash script mentioned in the issue description. Though that changes the name, while I encountered it initially when changing the `additional_hostnames`, this has stopped happening.
I still have the initial project that I encountered it on showing up twice when I do `ddev list` even though it's stopped and I have no other projects with it's name. (I would like to share it, but the name could be a giveaway on what the project is about and I'm not sure I'm allowed to publicly share that).

To me it seems like the old project names aren't properly removed, as both my initial project that I encountered it on and the one I tested with here, appear several times in `ddev list` even after they have been stopped and the project removed (it only removed one of them from the list, not all).

To summarise:

- Am unable to replicate the original way I encountered this issue (by having an additional hostname with the same name as project)
- Replicating it using the bash script in issue description does indeed replicate the issue.
- Using `ddev stop --remove-data` properly removes one of the entries from `ddev list` but the rest are still left on the list.

I hope this information helps.

**Comment by Graloth** [comment]
I just re-read your comment @rfay and saw that you're referencing the .ddev directory in my user directory :) my bad.

I would like to share it, but it contains even more references to projects that I cannot share, so I have removed them from the file, one of them was the one where I encountered the issue initially.

```yaml
APIVersion: v1.11.2
omit_containers: []
instrumentation_opt_in: false
router_bind_all_interfaces: false
last_used_version: v1.11.2
mkcert_caroot: /Users/dsj/Library/Application Support/mkcert
project_info:
  donationform:
    approot: /Users/dsj/Projects/Repositories/donationform
  foo-bar:
    approot: /Users/dsj/Projects/Repositories/foo-bar
  foo-example:
    approot: /Users/dsj/Projects/Repositories/foo-bar
  graloth:
    approot: /Users/dsj/Projects/Repositories/graloth

# You can turn off usage of the dba (phpmyadmin) container and/or
# ddev-ssh-agent containers with
# omit_containers[\"dba\", \"ddev-ssh-agent\"]
# and you can opt in or out of sending instrumentation the ddev developers with
# instrumentation_opt_in: true # or false
#
# instrumentation_user: <your_username> # can be used to give ddev specific info about who you are
# developer_mode: true # (defaults to false) is not used widely at this time.
# router_bind_all_interfaces: false  # (defaults to false)
#    If true, ddev-router will bind http/s, PHPMyAdmin, and MailHog ports on all
#    network interfaces instead of just localhost, so others on your local network can
#    access those ports. Note that this exposes the PHPMyAdmin and MailHog ports as well, which
#    can be a major security issue, so choose wisely. Consider omit_containers[dba] to avoid
#    exposing PHPMyAdmin.
```

What I did notice was that the initial project I had showing double, was because when I generated the config for the project it took the folder name as project name and I changed that afterwards (while messing with the additional hostnames), I think that might have caused my initial issue, since it appeared twice in the file above, one with the folder name and one with the actual project name.

**Comment by rfay** [comment]
That would definitely do it. ddev does try its best to prevent this, but that definitely causes the behavior.

**Comment by wolffc** [comment]
same here: 
```
APIVersion: v1.11.2
omit_containers: []
instrumentation_opt_in: true
router_bind_all_interfaces: false
last_used_version: v1.6.0
mkcert_caroot: C:\Users\cwolff\AppData\Local\mkcert
project_info:
  autotranslator-console:
    approot: C:\src\autotranslator-console
  bar-example:
    approot: C:\src\ddev-namechange
  cockpit-aer-de:
    approot: C:\src\ddev-cockpit
  cockpit-best-de:
    approot: C:\src\ddev-cockpit
  cockpit-conso-de:
    approot: C:\src\ddev-cockpit
  cockpit-dtt-de:
    approot: C:\src\ddev-cockpit
  cockpit-picasso-com:
    approot: C:\src\ddev-cockpit
  foo-bar:
    approot: C:\src\ddev-namechange
  foo-example:
    approot: C:\src\ddev-namechange
```
cockpit-* are my initial polymorph project while foo/bar are the test project from above

**Comment by rfay** [comment]
I see you've been using the same directory for multiple projects. It may be that ddev's defensive work is all about preventing the same project name from being reused, but it's not defensive about having the same directory reused. Simple workaround of course: Delete project before renaming it.

**Comment by wolffc** [comment]
actuall `ddev delete` is not wat we want.

we have a typo3 project which is is deployed to 13 servers sytems. (all the same code base) 
this project includes mutliple services. for each system 4 addtional db container,  solr, beanstalkd.. etc.

what we want is a fast way to switch from site A to B while. this includes restoring the data of the services if present. changing the project names gives us the ability (because the containers get different names) 

as a workaround our "switch-site script" now uses `ddev stop --unlist` to stop the project list from ever growing but avoding deltetion of the project data.

**Comment by rfay** [comment]
I think you'd be better off with projects of different names, I guess I don't completely understand your use-case. I'd recommend having projects (and probably repos) that have different names.

One thing that might be useful (not sure) if you have to do this is a "config.name.yaml" which just has `name: <specialnamehere>`, then you could overwrite the name without changing the config.yaml

**Comment by wolffc** [comment]
our project setup is like this:
- 1x Codebase which deploys ~ 15 websites different companies 
- 15 Development DB dumps (nightly, scrubbed from userdata)
- additional every DB needs an an indipentend set of Services.  so website 1 does not use the solr core of website 2 etc..
- we have a `switch-site script` which activates the Correct Configuration set. 

but as the configuration contains references to ID values of database reccords we can only use the Config set with the correct database. 

as a developer I don't want to dublicate the same git repository 15 times. to have a working test setup. thats why we create a script "switches from one site to another"

we do so by using the some "features" of ddev. if we change the project name, ddev creates containers with the site name in it. (which is great because wie dont have to deal with the process of backing up a project state. but rather say `ddev stop && change name && dev start` and ddev will start up with a set of independant containers. so ddev actually works great for that purporse.

and the dublication of "project names" in `global_config.yaml` is just a minor issue for us. overall the workflow seems to be working very well.

and to focus on the Project name issue:
i think the following rules would make sense for ddev to enforce:
a) you are not allows to use the same project name twice
b) you can only have one project per path. (unless ddev supports this)

so if i change the project name i would exprect ddev on ddev start to look if there are other proejct defined for this path, and remove non matching from `global_config.yaml`

**Comment by rfay** [comment]
Thanks for explaining that. Pretty fancy! I imagine you'll be able to get things to work by using `ddev stop --unlist` at the right place, etc.

ddev already tries really hard to enforce "you are not allows to use the same project name twice", but I think not hard enough to enforce "you can only have one project per path"

**Comment by chellman** [comment]
I've experienced the duplicate project name phenomenon as well, a little differently. I've created a project the regular way with `ddev config`, started it and worked with it for a while, then realized I had a typo in the name.  So I update it, `ddev restart`, and everything looks fine until I `ddev list` and see that there are now two projects with apparently the same name.

The global config file has the two projects listed with the same approot, but different names.  The tricky thing is that `ddev list` shows the **directory name** in its NAME column instead of the configured project name. Once I found that I could get the project name from the global config file and `ddev delete <duplicate_project>`, everything was fine.

**Comment by rfay** [comment]
I see two things that might be more defensive about this:

* `ddev start` should refuse to start if the name is different and the directory is already in use. That will prevent most of these. In the error message it should hint the removal technique (`ddev stop --unlist <projectname>`)
* `ddev list` can maybe show the actual name shown in global_config.yaml (as the key) instead of the name in the container label.  That would make it less confusing.
* Review the errors that get shown on duplicate name, so they provide info on how to work around (`ddev stop --unlist`)
